### PR TITLE
COMP: Install python packages using PyPI wheels

### PR DESCRIPTION
**Pull request by jamesobutler**
This updates the build system to install python packages using the official Wheels from PyPI instead of creating Eggs. This change is possible as Slicer now uses the same compiler as is used to generate the PyPI wheels or is at least binary compatible. This PR maintains the version used for each python package.  

The biggest change is probably getting the Numpy wheel from PyPI instead of building it from source and applying patches. The correct Numpy wheel based on platform/python version etc will be handled by the fetch process from PyPI.

### CODE REVIEW

**Comment by jcfr** [comment]
:pray: 

Thanks so much for the contribution, I will review in more details asap and post feedback.

After a cursory check, it looks good.

**Comment by pieper** [comment]
I also took a quick look and like it.  If Jc is good with it so am I.  Looks like the kind of change we can just merge and then tweak if anything comes up on specific platforms.

**Comment by sjh26** [comment]
👍

File: SuperBuild.cmake [hunk 1]
```diff
@@ -118,28 +118,25 @@ if(Slicer_BUILD_DICOM_SUPPORT)
...
-    python-couchdb
-    python-GitPython
-    python-six
+    python-extension-manager-requirements
     )
   if(Slicer_USE_PYTHONQT_WITH_OPENSSL OR Slicer_USE_SYSTEM_python)
     # python-PyGithub requires SSL support in Python
-    list(APPEND Slicer_DEPENDENCIES python-PyGithub)
+    list(APPEND Slicer_DEPENDENCIES
+      python-extension-manager-other-requirements
```
Comment (jcfr) [inline SuperBuild.cmake]: I suggest to rename this to `python-extension-manager-ssl-requirements`

File: SuperBuild/External_NUMPY.cmake [hunk 2]
```diff
@@ -1,164 +0,0 @@
...
-      python_DIR
-      PYTHON_SITE_PACKAGES_SUBDIR
-      )
-    if("${${varname}}" STREQUAL "")
-      message(FATAL_ERROR "${varname} CMake variable is expected to be set")
-    endif()
-  endforeach()
-
-  #-----------------------------------------------------------------------------
-  # Launcher setting specific to build tree
```
Comment (jcfr) [inline SuperBuild/External_NUMPY.cmake]: I still think the launcher settings should be updated in the current branch.

**Comment by jamesobutler** [comment]
Running tests locally as well right now. Thanks for the reviews!

**Comment by jcfr** [comment]
Trying to configure after deleting `python-*` from the top level directories returns the following error:

```
-- SuperBuild - First pass
CMake Error at CMake/ExternalProjectDependency.cmake:858 (message):
  Can't find External_python-nose.cmake
Call Stack (most recent call first):
  SuperBuild/External_python-numpy.cmake:17 (ExternalProject_Include_Dependencies)
  CMake/ExternalProjectDependency.cmake:842 (include)
  SuperBuild/External_python-dicom-requirements.cmake:19 (ExternalProject_Include_Dependencies)
  CMake/ExternalProjectDependency.cmake:842 (include)
  SuperBuild.cmake:473 (ExternalProject_Include_Dependencies)
  CMakeLists.txt:675 (include)


-- Configuring incomplete, errors occurred!
See also "/home/jcfr/Projects/Slicer-Qt5-VTK8-Release/CMakeFiles/CMakeOutput.log".

```

**Comment by jcfr** [comment]
I remove the extra `python-nose` dependency from `External_python-numpy.cmake` and it addressed the issue.

**Comment by jcfr** [comment]
It now fails with:

```
In --require-hashes mode, all requirements must have their versions pinned with ==. These do not:
    six from https://files.pythonhosted.org/packages/73/fb/00a976f728d0d1fecfe898238ce23f502a721c0ac0ecfedb80e0d88c64e9/six-1.12.0-py2.py3-none-any.whl#sha256=3350809f0555b11f552448330d0b52d5f24c91a322ea4a15ef22629740f3761c (from packaging==19.0->-r /home/jcfr/Projects/Slicer-Qt5-VTK8-Release/python-pythonqt-requirements-requirements.txt (line 2))
```

This was addressed adding a six to the pythonqt requirements

**Comment by jcfr** [comment]
Next failure were missing hash for nose and numpy, this was also easily fixed

**Comment by jcfr** [comment]
Remaining issues have been addressed. Closing, changes integrated as [r28540](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=28540)

**Comment by jamesobutler** [comment]
@jcfr Sorry I hadn't gotten to fixing the necessary changes.

Additional changes are needed:
Hash checks need to be removed from Numpy as there are different hashes for different platforms. Therefore `--require-hashes` needs to be removed and hashes removed from `nose` and `numpy`.
https://github.com/Slicer/Slicer/blob/ff86c6db83824c207cd7eb41d0ef9fcb8a15aaa8/SuperBuild/External_python-numpy.cmake#L8-L9

**Comment by jamesobutler** [comment]
Or do some cmake logic and get the Numpy hash for the specific python version and platform used if we always want to include hash checks.

**Comment by jcfr** [comment]
> for the specific python version and platform used if we always want to include hash checks.

Based on the error I got, specifying the hash is required

> do some cmake logic and get the Numpy hash for the specific python version and platform

:+1: 

Could you work on a fix for that ? 

The code could be changed to something like this:

```
if(WIN32)
  set(numpy_wheel_hash "")
elseif(APPLE)
  set(numpy_wheel_hash "")
else() # linux
  set(numpy_wheel_hash "23cc40313036cffd5d1873ef3ce2e949bdee0646c5d6f375bf7ee4f368db2511")
endif()
set(requirements_file ${CMAKE_BINARY_DIR}/${proj}-requirements.txt)
file(WRITE ${requirements_file} "
nose==1.3.7 --hash=sha256:9ff7c6cc443f8c51994b34a667bbcf45afd6d945be7477b52e97516fd17c53ac  # needed for NumPy unit tests
numpy==1.16.2 --hash=sha256:${numpy_wheel_hash} # auto pick platform
")
```

**Comment by jamesobutler** [comment]
> Could you work on a fix for that ?

👍 Will issue a PR with the change here soon

**Comment by jcfr** [comment]
**UPDATE**, it is possible to specify multiple hashes. See https://pip.pypa.io/en/stable/reference/pip_install/#hash-checking-mode

```
set(requirements_file ${CMAKE_BINARY_DIR}/${proj}-requirements.txt)
file(WRITE ${requirements_file} [===[
nose==1.3.7 --hash=sha256:9ff7c6cc443f8c51994b34a667bbcf45afd6d945be7477b52e97516fd17c53ac  # needed for NumPy unit tests
numpy==1.16.2 --hash=sha256:23cc40313036cffd5d1873ef3ce2e949bdee0646c5d6f375bf7ee4f368db2511 \
              --hash=sha256:23cc40313036cffd5d1873ef3ce2e949bdee0646c5d6f375bf7ee4f368db2511 \
              --hash=sha256:23cc40313036cffd5d1873ef3ce2e949bdee0646c5d6f375bf7ee4f368db2511
]===])
```

**Comment by pieper** [comment]
FYI I did a toplevel rebuild on mac with the and everything looks fine. Nice work 👍

**Pull request closed without merging**